src/cascadia/TerminalSettingsModel/ActionArgsMagic.h

```diff
@@ -54,50 +54,87 @@ struct InitListPlaceholder
 // expanded. Pretty critical for tracking down extraneous commas, etc.
 
 // Property definitions, and JSON keys
-#define DECLARE_ARGS(type, name, jsonKey, required, ...)    \
-    static constexpr std::string_view name##Key{ jsonKey }; \
+#define DECLARE_ARGS(type, name, jsonKey, required, tag, ...) \
+    static constexpr std::string_view name##Key{ jsonKey };   \
     ACTION_ARG(type, name, ##__VA_ARGS__);
 
 // Parameters to the non-default ctor
-#define CTOR_PARAMS(type, name, jsonKey, required, ...) \
+#define CTOR_PARAMS(type, name, jsonKey, required, tag, ...) \
     const type &name##Param,
 
 // initializers in the ctor
-#define CTOR_INIT(type, name, jsonKey, required, ...) \
+#define CTOR_INIT(type, name, jsonKey, required, tag, ...) \
     _##name{ name##Param },
 
+#define ARG_DESC_STRINGIFY2(x) #x
+#define ARG_DESC_STRINGIFY(x) ARG_DESC_STRINGIFY2(x)
+#define ARG_DESC_WIDEN2(x) L##x
+#define ARG_DESC_WIDEN(x) ARG_DESC_WIDEN2(x)
+#define LOCALIZED_NAME(name) ARG_DESC_WIDEN(ARG_DESC_STRINGIFY(name##Localized))
+
+// append this argument's description to the internal vector
+#define APPEND_ARG_DESCRIPTION(type, name, jsonKey, required, tag, ...) \
+    _argDescriptors.push_back({ RS_(LOCALIZED_NAME(name)), L## #type, std::wstring_view(L## #required) != L"false", tag });
+
 // check each property in the Equals() method. You'll note there's a stray
 // `true` in the definition of Equals() below, that's to deal with trailing
 // commas
-#define EQUALS_ARGS(type, name, jsonKey, required, ...) \
+#define EQUALS_ARGS(type, name, jsonKey, required, tag, ...) \
     &&(otherAsUs->_##name == _##name)
 
+// getter and setter for each property by index
+#define GET_ARG_BY_INDEX(type, name, jsonKey, required, tag, ...)    \
+    if (index == curIndex++)                                         \

```

Comment:
I think the `__COUNTER__` macro would be a good fit here.

```cpp
#define X_MACRO_INDEX_BASE() \
    constexpr auto X_MACRO_INDEXED_BASE__ = __COUNTER__ - 1

#define X_MACRO_INDEX() \
    (__COUNTER__ - X_MACRO_INDEXED_BASE__)

#define APPLY(foo, bar)   \
    case X_MACRO_INDEX(): \
        /* ... */         \
        break;

X_MACRO_INDEX_BASE();
switch (index) {
    SET_TAB_COLOR_ARGS(APPLY)
default:
    break;
}
```

Reply:
i'm in support of this one